cli/lsp/registries.rs

```diff
@@ -0,0 +1,823 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+use super::language_server;
+use super::path_to_regex::parse;
+use super::path_to_regex::string_to_regex;
+use super::path_to_regex::Compiler;
+use super::path_to_regex::Key;
+use super::path_to_regex::MatchResult;
+use super::path_to_regex::Matcher;
+use super::path_to_regex::StringOrNumber;
+use super::path_to_regex::StringOrVec;
+use super::path_to_regex::Token;
+
+use crate::deno_dir;
+use crate::file_fetcher::CacheSetting;
+use crate::file_fetcher::FileFetcher;
+use crate::http_cache::HttpCache;
+
+use deno_core::error::anyhow;
+use deno_core::error::AnyError;
+use deno_core::error::Context;
+use deno_core::resolve_url;
+use deno_core::serde::Deserialize;
+use deno_core::serde_json;
+use deno_core::serde_json::json;
+use deno_core::url::Position;
+use deno_core::url::Url;
+use deno_core::ModuleSpecifier;
+use deno_runtime::permissions::Permissions;
+use log::error;
+use lspower::lsp;
+use regex::Regex;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::path::Path;
+
+const CONFIG_PATH: &str = "/.well-known/deno-import-intellisense.json";
+const COMPONENT: &percent_encoding::AsciiSet = &percent_encoding::CONTROLS
+  .add(b' ')
+  .add(b'"')
+  .add(b'#')
+  .add(b'<')
+  .add(b'>')
+  .add(b'?')
+  .add(b'`')
+  .add(b'{')
+  .add(b'}')
+  .add(b'/')
+  .add(b':')
+  .add(b';')
+  .add(b'=')
+  .add(b'@')
+  .add(b'[')
+  .add(b'\\')
+  .add(b']')
+  .add(b'^')
+  .add(b'|')
+  .add(b'$')
+  .add(b'&')
+  .add(b'+')
+  .add(b',');
+
+lazy_static::lazy_static! {
+  static ref REPLACEMENT_VARIABLE_RE: Regex =
+    Regex::new(r"\$\{\{?(\w+)\}?\}").unwrap();
+}
+
+fn base_url(url: &Url) -> Result<Url, AnyError> {
```

Comment:
Can you use `Url::origin` instead?

Reply:
Url::origin is a tuple that is really hard to work with, especially to transform it back into a Url.

Reply:
You can do `Url::origin(&url).ascii_serialization()`. This will give you a string. It matches the output of `URL.origin` in JS. Alternatively you can do `url::quirks::origin(&url)`. This is the same as `Url::origin(&url).ascii_serialization()` under the hood though.

Reply:
String != Url... 😄   There is also `url[..Position:BeforePath]` if we wanted a string too...  But we need a valid base URL.  Could re-parse the string too, but I don't see an advantage to that.

Reply:
Sure, but what we want is the origin of the URL not an arbitrary base url. I don't see why we can't use the provided API for this. Origins are not module specifiers or URLs. They are opaque strings describing the origin of a URL (and happens to also be a valid URL). For example `xn--lt9h.deno.dev` and `🦕.deno.dev` should be the same origin. I think `ModuleRegistry#origins` should be changed to `HashMap<String, Vec<RegistryConfiguration>>`. See the patch below:

```patch